File: packages/contracts/test/remote-evidence.test.mjs
import assert from "node:assert/strict";
import test from "node:test";

import {
  assertExecutionCommand,
  providerInputAttestationDigest,
  providerObservationDigest,
  remoteCIObservationReceiptDigest,
  remoteCommitObservationDigest,
  remoteInputAttestationDigest,
  remoteInputEvidenceDigest,
  remoteProviderBindingDigest,
  remoteProviderBindingRevocationDigest
} from "../src/execution-validation.mjs";

const at = "2026-09-02T12:00:00.000Z";
const digest = (character) => character.repeat(64);
const sha1 = (character) => character.repeat(40);

function binding(origin = "https://provider.example") {
  const value = {
    version: 1,
    providerBindingId: "provider_example001",
    teamId: "team_example001",
    repositoryId: "repo_example001",
    providerOrigin: origin,
    providerRepositoryId: "owner/repository",
    ciChecks: [{
      checkKey: "unit",
      profileId: "profile_example001",
      profileRevision: 1,
      profileDigest: digest("a")
    }],
    createdByMemberId: "member_example001",
    bindingDigest: digest("0"),
    createdAt: at
  };
  value.bindingDigest = remoteProviderBindingDigest(value);
  return value;
}

function providerCommit() {
  const value = {
    version: 1,
    operationId: "op_remote_commit0001",
    observationId: "observation_commit0001",
    providerRepositoryId: "owner/repository",
    objectFormat: "sha1",
    baseCommit: sha1("1"),
    commit: sha1("2"),
    tree: sha1("3"),
    bundleDigest: digest("b"),
    bundleByteLength: 100,
    pullRequest: {
      number: 17,
      headRef: "refs/heads/change",
      baseRef: "refs/heads/main"
    },
    providerObservationDigest: digest("0"),
    observedAt: at
  };
  value.providerObservationDigest = providerObservationDigest(value);
  return value;
}

function providerCI() {
  const value = {
    version: 1,
    operationId: "op_remote_ci0000001",
    observationId: "observation_ci00000001",
    providerRepositoryId: "owner/repository",
    checkKey: "unit",
    attempt: 1,
    commit: sha1("2"),
    tree: sha1("3"),
    outcome: "passed",
    providerObservationDigest: digest("0"),
    observedAt: at
  };
  value.providerObservationDigest = providerObservationDigest(value);
  return value;
}

function providerInputAttestation() {
  const value = {
    version: 1,
    operationId: "op_remote_inputs00001",
    attestationId: "attestation_remoteinputs0001",
    providerRepositoryId: "owner/repository",
    nodeKey: "Build",
    commit: sha1("2"),
    tree: sha1("3"),
    inputs: [{
      adoptionId: "adoption_inputsource0001",
      adoptionDigest: digest("a"),
      reuseInput: {
        inputSlot: "source",
        producer: {
          kind: "adopted_evidence",
          edge: {
            edgeKey: "PrepareBuild",
            fromNodeKey: "Prepare",
            toNodeKey: "Build",
            gate: "verified_output",
            bindings: [{ outputSlot: "patch", inputSlot: "source" }]
          },
          sourceEvidenceId: "source_inputsource0001",
          sourceDigest: digest("b"),
          proofSetDigest: digest("c")
        },
        artifact: { contentDigest: digest("d"), kind: "patch" }
      }
    }],
    remoteInputEvidenceDigest: digest("0"),
    providerAttestationDigest: digest("0"),
    attestedAt: at
  };
  value.remoteInputEvidenceDigest = remoteInputEvidenceDigest(value.inputs);
  value.providerAttestationDigest = providerInputAttestationDigest(value);
  return value;
}

test("remote provider binding is metadata-only, ordered and digest-bound", () => {
  for (const origin of ["https://provider.example", "http://127.0.0.1:4317", "http://[::1]:4317"]) {
    assert.doesNotThrow(() => assertExecutionCommand("remoteProviderBinding", binding(origin)));
  }
  for (const origin of [
    "http://provider.example", "https://token@provider.example",
    "https://provider.example/path", "https://provider.example?token=secret"
  ]) {
    assert.throws(() => assertExecutionCommand("remoteProviderBinding", binding(origin)));
  }
  const unordered = binding();
  unordered.ciChecks = [
    { checkKey: "z", profileId: "profile_example002", profileRevision: 1, profileDigest: digest("b") },
    ...unordered.ciChecks
  ];
  unordered.bindingDigest = remoteProviderBindingDigest(unordered);
  assert.throws(() => assertExecutionCommand("remoteProviderBinding", unordered),
    /REMOTE_PROVIDER_CHECK_ORDER/u);
  assert.throws(() => assertExecutionCommand("remoteProviderBinding", {
    ...binding(), credential: "forbidden"
  }), /PLAN_SCHEMA_INVALID/u);
  assert.throws(() => assertExecutionCommand("remoteProviderBinding", {
    ...binding(), bindingDigest: digest("f")
  }), /REMOTE_PROVIDER_BINDING_DIGEST_MISMATCH/u);
});

test("provider binding revocation is immutable-payload digest-bound", () => {
  const value = {
    version: 1,
    operationId: "op_provider_revoke0001",
    providerBindingId: "provider_example001",
    expectedBindingDigest: binding().bindingDigest,
    revokedByMemberId: "member_example001",
    reason: "Owner withdrew remote observation authority",
    revocationDigest: digest("0"),
    revokedAt: at
  };
  value.revocationDigest = remoteProviderBindingRevocationDigest(value);
  assert.doesNotThrow(() => assertExecutionCommand("remoteProviderBindingRevocation", value));
  assert.throws(() => assertExecutionCommand("remoteProviderBindingRevocation", {
    ...value, expectedBindingDigest: digest("f")
  }), /REMOTE_PROVIDER_REVOCATION_DIGEST_MISMATCH/u);
});

test("provider and retained commit observations bind exact Git and Artifact identity", () => {
  const external = providerCommit();
  assert.doesNotThrow(() => assertExecutionCommand("providerCommitObservation", external));
  assert.throws(() => assertExecutionCommand("providerCommitObservation", {
    ...external, tree: digest("3"),
    providerObservationDigest: providerObservationDigest({ ...external, tree: digest("3") })
  }), /REMOTE_EVIDENCE_OBJECT_FORMAT_MISMATCH/u);
  assert.throws(() => assertExecutionCommand("providerCommitObservation", {
    ...external, providerObservationDigest: digest("f")
  }), /REMOTE_PROVIDER_OBSERVATION_DIGEST_MISMATCH/u);

  const retained = {
    version: 1,
    operationId: external.operationId,
    providerBindingId: "provider_example001",
    repositoryId: "repo_example001",
    providerRepositoryId: external.providerRepositoryId,
    taskId: "task_example001",
    observationId: external.observationId,
    objectFormat: external.objectFormat,
    baseCommit: external.baseCommit,
    commit: external.commit,
    tree: external.tree,
    inputDigest: digest("d"),
    bundleArtifactId: "artifact_bundle0001",
    bundleDigest: external.bundleDigest,
    bundleByteLength: external.bundleByteLength,
    patchArtifactId: "artifact_patch00001",
    patchArtifactRevision: 1,
    patchOutputSlot: "patch",
    patchDigest: digest("e"),
    patchByteLength: 20,
    pullRequest: external.pullRequest,
    providerObservationDigest: external.providerObservationDigest,
    observationDigest: digest("0"),
    observedAt: at
  };
  retained.observationDigest = remoteCommitObservationDigest(retained);
  assert.doesNotThrow(() => assertExecutionCommand("remoteCommitObservation", retained));
  assert.throws(() => assertExecutionCommand("remoteCommitObservation", {
    ...retained, patchArtifactId: "artifact_other0001"
  }), /REMOTE_COMMIT_OBSERVATION_DIGEST_MISMATCH/u);
});

test("remote CI receipt is distinct from provider observation and exact-source bound", () => {
  const external = providerCI();
  assert.doesNotThrow(() => assertExecutionCommand("providerCIObservation", external));
  const receipt = {
    ...external,
    providerBindingId: "provider_example001",
    repositoryId: "repo_example001",
    sourceEvidenceId: "source_example0001",
    profileId: "profile_example001",
    profileRevision: 1,
    profileDigest: digest("a"),
    receiptDigest: digest("0")
  };
  receipt.receiptDigest = remoteCIObservationReceiptDigest(receipt);
  assert.doesNotThrow(() => assertExecutionCommand("remoteCIObservationReceipt", receipt));
  assert.throws(() => assertExecutionCommand("remoteCIObservationReceipt", {
    ...receipt, checkKey: "foreign"
  }), /REMOTE_CI_RECEIPT_DIGEST_MISMATCH/u);
  assert.throws(() => assertExecutionCommand("providerCIObservation", {
    ...external, callbackSecret: "forbidden"
  }), /PLAN_SCHEMA_INVALID/u);
});

test("remote input attestations separate adoption authority from logical equality", () => {
  const external = providerInputAttestation();
  assert.doesNotThrow(() =>
    assertExecutionCommand("providerInputAttestation", external));
  const retained = {
    ...external,
    providerBindingId: "provider_example001",
    repositoryId: "repo_example001",
    planId: "plan_example001",
    planRevision: 1,
    sourceEvidenceId: "source_remoteoutput0001",
    sourceDigest: digest("e"),
    sourceObservationId: "observation_commit0001",
    sourceObservationDigest: digest("f"),
    attestationDigest: digest("0")
  };
  retained.attestationDigest = remoteInputAttestationDigest(retained);
  assert.doesNotThrow(() =>
    assertExecutionCommand("remoteInputAttestation", retained));
  assert.equal(retained.remoteInputEvidenceDigest,
    remoteInputEvidenceDigest(retained.inputs));

  const secondAuthority = structuredClone(external);
  secondAuthority.inputs[0].adoptionId = "adoption_inputsource0002";
  secondAuthority.inputs[0].adoptionDigest = digest("9");
  secondAuthority.providerAttestationDigest =
    providerInputAttestationDigest(secondAuthority);
  assert.equal(secondAuthority.remoteInputEvidenceDigest,
    external.remoteInputEvidenceDigest);
  assert.notEqual(secondAuthority.providerAttestationDigest,
    external.providerAttestationDigest);

  for (const mutate of [
    (value) => { value.inputs[0].reuseInput.producer.sourceDigest = digest("1"); },
    (value) => { value.inputs[0].reuseInput.producer.proofSetDigest = digest("2"); },
    (value) => { value.inputs[0].reuseInput.artifact.contentDigest = digest("3"); },
    (value) => { value.inputs[0].reuseInput.producer.edge.toNodeKey = "Other"; },
    (value) => { value.tree = digest("4"); },
    (value) => { value.credential = "forbidden"; }
  ]) {
    const invalid = structuredClone(external);
    mutate(invalid);
    assert.throws(() =>
      assertExecutionCommand("providerInputAttestation", invalid));
  }

  const externalProducer = structuredClone(external);
  externalProducer.inputs[0].reuseInput.producer = {
    kind: "external_result",
    externalInput: {
      nodeKey: "Build",
      inputSlot: "source",
      sourceTaskId: "task_source0001",
      sourceResultId: "result_source0001",
      artifactId: "artifact_source0001",
      artifactRevision: 1,
      contentDigest: digest("d"),
      kind: "patch"
    },
    reviewOperationId: "op_review_source0001",
    reviewDigest: digest("5")
  };
  externalProducer.remoteInputEvidenceDigest =
    remoteInputEvidenceDigest(externalProducer.inputs);
  externalProducer.providerAttestationDigest =
    providerInputAttestationDigest(externalProducer);
  assert.throws(() =>
    assertExecutionCommand("providerInputAttestation", externalProducer),
  /REMOTE_INPUT_ATTESTATION_PRODUCER_INVALID/u);
});

File: packages/contracts/test/go/remote_evidence_test.go
package contracts_test

import (
	"encoding/json"
	"reflect"
	"strings"
	"testing"

	execution "convenewire.dev/contracts/generated/go/execution"
	runtimecontracts "convenewire.dev/contracts/generated/go/runtime"
)

func sealRemote(t *testing.T, value map[string]any, digestField string, excluded ...string) {
	t.Helper()
	value[digestField] = executionDigest(t, withoutFields(value, excluded...))
}

func TestRemoteEvidenceGoValidation(t *testing.T) {
	at := "2026-09-02T12:00:00.000Z"
	binding := map[string]any{
		"version": 1, "providerBindingId": "provider_example001",
		"teamId": "team_example001", "repositoryId": "repo_example001",
		"providerOrigin": "http://127.0.0.1:4317", "providerRepositoryId": "owner/repository",
		"ciChecks": []any{map[string]any{
			"checkKey": "unit", "profileId": "profile_example001",
			"profileRevision": 1, "profileDigest": strings.Repeat("a", 64),
		}},
		"createdByMemberId": "member_example001", "bindingDigest": strings.Repeat("0", 64),
		"createdAt": at,
	}
	sealRemote(t, binding, "bindingDigest", "bindingDigest", "createdAt")
	if err := runtimecontracts.ValidateExecutionCommand("remoteProviderBinding", encodeJSON(t, binding)); err != nil {
		t.Fatalf("valid remote binding rejected: %v", err)
	}
	badOrigin := decodeMap(t, encodeJSON(t, binding))
	badOrigin["providerOrigin"] = "http://provider.example"
	sealRemote(t, badOrigin, "bindingDigest", "bindingDigest", "createdAt")
	if runtimecontracts.ValidateExecutionCommand("remoteProviderBinding", encodeJSON(t, badOrigin)) == nil {
		t.Fatal("public cleartext provider origin accepted")
	}

	providerCommit := map[string]any{
		"version": 1, "operationId": "op_remote_commit0001",
		"observationId": "observation_commit0001", "providerRepositoryId": "owner/repository",
		"objectFormat": "sha1", "baseCommit": strings.Repeat("1", 40),
		"commit": strings.Repeat("2", 40), "tree": strings.Repeat("3", 40),
		"bundleDigest": strings.Repeat("b", 64), "bundleByteLength": 100,
		"pullRequest": nil, "providerObservationDigest": strings.Repeat("0", 64), "observedAt": at,
	}
	sealRemote(t, providerCommit, "providerObservationDigest", "providerObservationDigest")
	if err := runtimecontracts.ValidateExecutionCommand("providerCommitObservation", encodeJSON(t, providerCommit)); err != nil {
		t.Fatalf("valid provider commit rejected: %v", err)
	}
	wrongFormat := decodeMap(t, encodeJSON(t, providerCommit))
	wrongFormat["tree"] = strings.Repeat("3", 64)
	sealRemote(t, wrongFormat, "providerObservationDigest", "providerObservationDigest")
	if runtimecontracts.ValidateExecutionCommand("providerCommitObservation", encodeJSON(t, wrongFormat)) == nil {
		t.Fatal("wrong Git object format accepted")
	}

	providerCI := map[string]any{
		"version": 1, "operationId": "op_remote_ci0000001",
		"observationId": "observation_ci00000001", "providerRepositoryId": "owner/repository",
		"checkKey": "unit", "attempt": 1, "commit": strings.Repeat("2", 40),
		"tree": strings.Repeat("3", 40), "outcome": "passed",
		"providerObservationDigest": strings.Repeat("0", 64), "observedAt": at,
	}
	sealRemote(t, providerCI, "providerObservationDigest", "providerObservationDigest")
	if err := runtimecontracts.ValidateExecutionCommand("providerCIObservation", encodeJSON(t, providerCI)); err != nil {
		t.Fatalf("valid provider CI observation rejected: %v", err)
	}
	receipt := decodeMap(t, encodeJSON(t, providerCI))
	receipt["providerBindingId"] = "provider_example001"
	receipt["repositoryId"] = "repo_example001"
	receipt["sourceEvidenceId"] = "source_example0001"
	receipt["profileId"] = "profile_example001"
	receipt["profileRevision"] = 1
	receipt["profileDigest"] = strings.Repeat("a", 64)
	receipt["receiptDigest"] = strings.Repeat("0", 64)
	sealRemote(t, receipt, "receiptDigest", "receiptDigest")
	if err := runtimecontracts.ValidateExecutionCommand("remoteCIObservationReceipt", encodeJSON(t, receipt)); err != nil {
		t.Fatalf("valid remote CI receipt rejected: %v", err)
	}
	receipt["checkKey"] = "foreign"
	if runtimecontracts.ValidateExecutionCommand("remoteCIObservationReceipt", encodeJSON(t, receipt)) == nil {
		t.Fatal("tampered remote CI receipt accepted")
	}
}

func TestRemoteInputAttestationGoValidation(t *testing.T) {
	at := "2026-09-03T01:00:00.000Z"
	input := map[string]any{
		"adoptionId": "adoption_inputsource0001", "adoptionDigest": strings.Repeat("a", 64),
		"reuseInput": map[string]any{
			"inputSlot": "source",
			"producer": map[string]any{
				"kind": "adopted_evidence",
				"edge": map[string]any{
					"edgeKey": "PrepareBuild", "fromNodeKey": "Prepare", "toNodeKey": "Build",
					"gate": "verified_output", "bindings": []any{map[string]any{
						"outputSlot": "patch", "inputSlot": "source",
					}},
				},
				"sourceEvidenceId": "source_inputsource0001",
				"sourceDigest":     strings.Repeat("b", 64),
				"proofSetDigest":   strings.Repeat("c", 64),
			},
			"artifact": map[string]any{
				"contentDigest": strings.Repeat("d", 64), "kind": "patch",
			},
		},
	}
	provider := map[string]any{
		"version": 1, "operationId": "op_remote_inputs00001",
		"attestationId":        "attestation_remoteinputs0001",
		"providerRepositoryId": "owner/repository", "nodeKey": "Build",
		"commit": strings.Repeat("2", 40), "tree": strings.Repeat("3", 40),
		"inputs": []any{input}, "remoteInputEvidenceDigest": strings.Repeat("0", 64),
		"providerAttestationDigest": strings.Repeat("0", 64), "attestedAt": at,
	}
	provider["remoteInputEvidenceDigest"] = executionDigest(t, []any{input["reuseInput"]})
	sealRemote(t, provider, "providerAttestationDigest", "providerAttestationDigest")
	if err := runtimecontracts.ValidateExecutionCommand("providerInputAttestation", encodeJSON(t, provider)); err != nil {
		t.Fatalf("valid provider input attestation rejected: %v", err)
	}
	providerJSON := encodeJSON(t, provider)
	var typedProvider execution.ProviderInputAttestation
	if err := json.Unmarshal(providerJSON, &typedProvider); err != nil {
		t.Fatalf("generated provider type rejected valid JSON: %v", err)
	}
	typedProviderJSON, err := json.Marshal(typedProvider)
	if err != nil || !reflect.DeepEqual(decodeJSON(t, providerJSON), decodeJSON(t, typedProviderJSON)) {
		t.Fatal("generated provider type changed attestation JSON")
	}
	retained := decodeMap(t, encodeJSON(t, provider))
	retained["providerBindingId"] = "provider_example001"
	retained["repositoryId"] = "repo_example001"
	retained["planId"] = "plan_example001"
	retained["planRevision"] = 1
	retained["sourceEvidenceId"] = "source_remoteoutput0001"
	retained["sourceDigest"] = strings.Repeat("e", 64)
	retained["sourceObservationId"] = "observation_commit0001"
	retained["sourceObservationDigest"] = strings.Repeat("f", 64)
	retained["attestationDigest"] = strings.Repeat("0", 64)
	sealRemote(t, retained, "attestationDigest", "attestationDigest")
	if err := runtimecontracts.ValidateExecutionCommand("remoteInputAttestation", encodeJSON(t, retained)); err != nil {
		t.Fatalf("valid retained input attestation rejected: %v", err)
	}
	retainedJSON := encodeJSON(t, retained)
	var typedRetained execution.RemoteInputAttestation
	if err := json.Unmarshal(retainedJSON, &typedRetained); err != nil {
		t.Fatalf("generated retained type rejected valid JSON: %v", err)
	}
	typedRetainedJSON, err := json.Marshal(typedRetained)
	if err != nil || !reflect.DeepEqual(decodeJSON(t, retainedJSON), decodeJSON(t, typedRetainedJSON)) {
		t.Fatal("generated retained type changed attestation JSON")
	}
	tampered := decodeMap(t, encodeJSON(t, retained))
	tamperedInputs := tampered["inputs"].([]any)
	tamperedReuse := tamperedInputs[0].(map[string]any)["reuseInput"].(map[string]any)
	tamperedReuse["artifact"].(map[string]any)["contentDigest"] = strings.Repeat("9", 64)
	sealRemote(t, tampered, "attestationDigest", "attestationDigest")
	if runtimecontracts.ValidateExecutionCommand("remoteInputAttestation", encodeJSON(t, tampered)) == nil {
		t.Fatal("tampered logical input accepted")
	}
}

func TestProofControlCommandsHaveClosedGoValidation(t *testing.T) {
	digest := strings.Repeat("a", 64)
	remote := map[string]any{
		"operationId": "op_web_adopt00001", "providerBindingId": "provider_web00001",
		"planRevision": 1, "nodeKey": "Build", "expectedPlanDigest": digest,
		"expectedControlRevision": 2, "sourceEvidenceId": "source_web_remote01",
	}
	if err := runtimecontracts.ValidateExecutionCommand("remoteEvidenceAdoptionCommand", encodeJSON(t, remote)); err != nil {
		t.Fatalf("valid remote adoption command rejected: %v", err)
	}
	remote["credential"] = "forbidden"
	if runtimecontracts.ValidateExecutionCommand("remoteEvidenceAdoptionCommand", encodeJSON(t, remote)) == nil {
		t.Fatal("remote adoption command accepted an undeclared credential")
	}

	integration := map[string]any{
		"operationId": "op_web_integrate001", "candidateCommit": strings.Repeat("2", 40),
		"candidateTree": strings.Repeat("3", 40), "deadline": "2026-09-03T01:05:00.000Z",
		"inputDigest": digest, "materializationDigest": strings.Repeat("b", 64),
		"nodeKey": "Build", "planId": "plan_web0000001", "planRevision": 1,
		"target": map[string]any{
			"repositoryId": "repo_web00001", "targetRef": "refs/heads/main",
			"expectedCommit": strings.Repeat("1", 40),
		},
		"verificationReceipts": []any{map[string]any{
			"receiptDigest":  strings.Repeat("c", 64),
			"verificationId": "verification_web00001",
		}},
	}
	if err := runtimecontracts.ValidateExecutionCommand("integrationApprovalCommand", encodeJSON(t, integration)); err != nil {
		t.Fatalf("valid integration approval command rejected: %v", err)
	}
}
